README.md
# vue-countdown

> Simple Countdown component for [Vue.js](https://vuejs.org).

* [Example](https://hebertsilva.github.io/vue-countdown/)

## Project setup

In browser:

```html
<script src="/path/to/vue.js"></script>
<script src="/path/to/vue-countdown.js"></script>
<script>Vue.component(VueCountdown.name, VueCountdown);</script>
```

### Usage

```js
import Vue from 'vue'
import VueCountdown from '@hebertsilva/vue-countdown'
Vue.component(VueCountdown.name, VueCountdown)
```

```html
<countdown :time="20" :onComplete="setFunctionOnComplete" :labels="['Days', 'Hours', 'Minutes', 'Seconds']" />
<!-- 
    <ol>
        <li>
            <span class="time">00</span>
            <span class="label">Days</span>
        </li>
        <li>
            <span class="time">21</span>
            <span class="label">Horus</span>
        </li>
        <li>
            <span class="time">19</span>
            <span class="label">Minutes</span>
        </li>
        <li>
            <span class="time">47</span>
            <span class="label">Seconds</span>
        </li>
    </ol>
 -->
```

## Props

### time

- Type: `Number` *// Minutes*
- Default: `0`

### labels
- Type: `Array`
- Default: `['Days', 'Hours', 'Minutes', 'Seconds']`

### onComplete
- Type: `Function`
- Default: `null`

## License

[MIT](https://opensource.org/licenses/MIT) © Hebert Silva

CHANGELOG.md
## 1.0.1 (Jul 29, 2019)
- Fix name project for public
- Fix add urls repository

## 1.0.0 (Jul 27, 2019)
- Props in `labels` : `['Days', 'Hours', 'Minutes', 'Seconds']`
- Props after the time is up: `onComplete()`

test/index.js
import Vue from 'vue'
import VueCountdown from '../src/index.js'

Vue.component(VueCountdown.name, VueCountdown)


test/Countdown.spec.js
describe('Countdown', () => {
  it('should progress only when it is counting the time is 20', () => {
    new Vue({
      template: '<countdown ref="countdown" :time="20" />',
    }).$mount()
  })

  it('should event onComplete when the time is 0', (done) => {
    new Vue({
      methods: {
        handleCountdownEnd() {
          done()
        }
      },
      template: '<countdown ref="countdown" :time="0" :onComplete="handleCountdownEnd()" />'
    }).$mount()
  })

  it('should event not visible labels when the component', () => {
    new Vue({
      template: '<countdown ref="countdown" :time="0" :labels="[]" />'
    }).$mount()
  })

  it('should event has visible labels when the component', () => {
    new Vue({
      template: `<countdown ref="countdown" :time="0" :labels="['Days', 'Hours', 'Minutes', 'Seconds']" />`
    }).$mount()
  })
})


src/index.js
export default {
  name: "countdown",
  props: {
    /**
     * Starts the countdown automatically when initialized.
     */
    time: {
      type: Number,
      default: 0
    },

    /**
     * Starts the countdown labels for language defined
     */
    labels: {
      type: Array,
      default() {
        return ["Days", "Hours", "Minutes", "Seconds"]
      }
    },

    /**
     * Emit event the countdown has completed
     */
    onComplete: {
      type: Function,
      default() {
        return null
      }
    }
  },
  data() {
    return {
      /**
       * It is counting down.
       * @type {Object}
       */
      countdown: {
        days: 0,
        hours: 0,
        minutes: 0,
        seconds: 0
      },

      /**
       * It is counting down.
       * @type {Boolean}
       */
      visible: true
    }
  },
  /**
   * Parse time in new Date() for time defined
   */
  computed: {
    parseTime() {
      return new Date(
        Date.parse(new Date()) + 1 * 1 * 1 * (this.time * 60) * 1000
      )
    }
  },
  beforeMount() {
    const time = this.getTimeRemaining(this.parseTime)

    if (time) {
      this.countdown = time
    }
  },
  mounted() {
    this.startCountdown()
  },
  render(createElement) {
    const strings = ["days", "hours", "minutes", "seconds"]

    const childrens = strings.map((v, i) => {
      return createElement("li", [
        createElement(
          "span",
          { class: "time" },
          this.labelTime(this.countdown[v])
        ),

        this.labels.length
          ? createElement("span", { class: "label" }, this.labels[i])
          : ""
      ])
    })

    return createElement("ol", childrens)
  },
  methods: {
    getTimeRemaining(time) {
      /**
       * Convert time in counter
       */
      const total = Date.parse(time) - Date.parse(new Date())
      const days = Math.floor(total / (1000 * 60 * 60 * 24))
      const hours = Math.floor((total / (1000 * 60 * 60)) % 24)
      const minutes = Math.floor((total / 1000 / 60) % 60)
      const seconds = Math.floor((total / 1000) % 60)

      return {
        total: total,
        days: days,
        hours: hours,
        minutes: minutes,
        seconds: seconds
      }
    },
    startCountdown() {
      /**
       * Start countdown in interval one seconds
       */
      const interval = setInterval(() => {
        const time = this.getTimeRemaining(this.parseTime)

        if (time) {
          this.countdown = time
        }

        if (time.total <= 0) {
          this.stopCountdown(interval)
          this.visible = false

          if (typeof this.onComplete === "function") {
            this.onComplete()
          }
        }
      }, 1000)
    },
    stopCountdown(interval) {
      /**
       * Stop countdown
       */
      clearInterval(interval)
    },
    labelTime(time) {
      /**
       * Convert unitiis time in string "00"
       */
      return time >= 0 ? ("0" + time).slice(-2) : "00"
    }
  }
}


karma.conf.js
const puppeteer = require('puppeteer')
const rollupConfig = require('./rollup.config');

process.env.CHROME_BIN = puppeteer.executablePath()
process.env.NODE_ENV = 'test'

module.exports = config => {
  config.set({
    autoWatch: false,
    browsers: ['ChromeHeadless'],
    coverageIstanbulReporter: {
      reports: ['html', 'lcovonly', 'text-summary'],
    },
    files: [
      'node_modules/vue/dist/vue.js',
      'test/index.js',
      'test/*.spec.js',
    ],
    frameworks: ['mocha', 'chai'],
    preprocessors: {
      'test/index.js': ['rollup'],
      'test/*.spec.js': ['rollup'],
    },
    reporters: ['mocha', 'coverage-istanbul'],
    rollupPreprocessor: {
      plugins: rollupConfig.plugins,
      output: {
        format: 'iife',
        name: 'Anonymous',
        sourcemap: 'inline',
      },
    },
    singleRun: true,
  })
}


assets/js/vue-countdown.js
/*!
 * vue-countdown v1.0.0
 * https://github.com/hebertsilva/vue-countdown#readme
 *
 * Copyright 2019-present Hebert Silva
 * Released under the MIT license
 *
 * Date: 2019-07-25T20:43:12.191Z
 */

(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = global || self, global.VueCountdown = factory());
}(this, function () { 'use strict';

  var index = {
    name: "countdown",
    props: {
      /**
       * Starts the countdown automatically when initialized.
       */
      time: {
        type: Number,
        default: 0
      },

      /**
       * Starts the countdown labels for language defined
       */
      labels: {
        type: Array,
        default: function _default() {
          return ["Days", "Hours", "Minutes", "Seconds"];
        }
      },

      /**
       * Emit event the countdown has completed
       */
      onComplete: {
        type: Function,
        default: function _default() {
          return null;
        }
      }
    },
    data: function data() {
      return {
        /**
         * It is counting down.
         * @type {Object}
         */
        countdown: {
          days: 0,
          hours: 0,
          minutes: 0,
          seconds: 0
        },

        /**
         * It is counting down.
         * @type {Boolean}
         */
        visible: true
      };
    },

    /**
     * Parse time in new Date() for time defined
     */
    computed: {
      parseTime: function parseTime() {
        return new Date(Date.parse(new Date()) + 1 * 1 * 1 * (this.time * 60) * 1000);
      }
    },
    beforeMount: function beforeMount() {
      var time = this.getTimeRemaining(this.parseTime);

      if (time) {
        this.countdown = time;
      }
    },
    mounted: function mounted() {
      this.startCountdown();
    },
    render: function render(createElement) {
      var _this = this;

      var strings = ["days", "hours", "minutes", "seconds"];
      var childrens = strings.map(function (v, i) {
        return createElement("li", [createElement("span", {
          class: "time"
        }, _this.labelTime(_this.countdown[v])), _this.labels.length ? createElement("span", {
          class: "label"
        }, _this.labels[i]) : ""]);
      });
      return createElement("ol", childrens);
    },
    methods: {
      getTimeRemaining: function getTimeRemaining(time) {
        /**
         * Convert time in counter
         */
        var total = Date.parse(time) - Date.parse(new Date());
        var days = Math.floor(total / (1000 * 60 * 60 * 24));
        var hours = Math.floor(total / (1000 * 60 * 60) % 24);
        var minutes = Math.floor(total / 1000 / 60 % 60);
        var seconds = Math.floor(total / 1000 % 60);
        return {
          total: total,
          days: days,
          hours: hours,
          minutes: minutes,
          seconds: seconds
        };
      },
      startCountdown: function startCountdown() {
        var _this2 = this;

        /**
         * Start countdown in interval one seconds
         */
        var interval = setInterval(function () {
          var time = _this2.getTimeRemaining(_this2.parseTime);

          if (time) {
            _this2.countdown = time;
          }

          if (time.total <= 0) {
            _this2.stopCountdown(interval);

            _this2.visible = false;

            if (typeof _this2.onComplete === "function") {
              _this2.onComplete();
            }
          }
        }, 1000);
      },
      stopCountdown: function stopCountdown(interval) {
        /**
         * Stop countdown
         */
        clearInterval(interval);
      },
      labelTime: function labelTime(time) {
        /**
         * Convert unitiis time in string "00"
         */
        return time >= 0 ? ("0" + time).slice(-2) : "00";
      }
    }
  };

  return index;

}));
